### PR TITLE
Implement iptable hook

### PR DESCRIPTION
**Pull request by WinterNis**

### CODE REVIEW

File: container/docker_test.go [hunk 1]
```diff
@@ -48,6 +53,29 @@ func (s *DockerContainerizerTestSuite) SetupTest() {
...
+			},
+		},
+	}
+
+	monkey.PatchInstanceMethod(
+		reflect.TypeOf(s.dc.Client),
+		"InspectContainer",
+		func(_ *docker.Client, id string) (*docker.Container, error) {
+			return s.container, nil
+		})
```
Comment (Devatoria) [inline container/docker_test.go]: Line jump between } and )

File: container/docker_test.go [hunk 2]
```diff
@@ -180,11 +213,49 @@ func (s *DockerContainerizerTestSuite) TestDockerContainerRemove() {
...
+	pid, err := s.dc.ContainerGetPID("abcdef1234")
 	assert.Nil(s.T(), err)
-	assert.Empty(s.T(), s.req.body)
+	assert.Equal(s.T(), s.container.State.Pid, pid)
+}
+
+// Check that returns ips in all networks
+// Check that returns empty map when container is in host mode
+func (s *DockerContainerizerTestSuite) TestDockerContainerGetIPs() {
+
```
Comment (Devatoria) [inline container/docker_test.go]: Empty line

File: container/docker_test.go [hunk 3]
```diff
@@ -180,11 +213,49 @@ func (s *DockerContainerizerTestSuite) TestDockerContainerRemove() {
...
+	// Check that returns ips in all networks
+	ips, err := s.dc.ContainerGetIPs("abcdef1234")
+	assert.Nil(s.T(), err)
+	assert.Equal(
+		s.T(),
+		map[string]net.IP{
+			"bridge":       net.ParseIP(s.container.NetworkSettings.Networks["bridge"].IPAddress),
+			"user_network": net.ParseIP(s.container.NetworkSettings.Networks["user_network"].IPAddress)},
+		ips)
```
Comment (Devatoria) [inline container/docker_test.go]: Line jump before )

File: container/docker_test.go [hunk 4]
```diff
@@ -180,11 +213,49 @@ func (s *DockerContainerizerTestSuite) TestDockerContainerRemove() {
...
+				State: docker.State{
+					Pid: 1024,
+				},
+				NetworkSettings: &docker.NetworkSettings{
+					Networks: map[string]docker.ContainerNetwork{
+						"host": docker.ContainerNetwork{},
+					},
+				},
+			}, nil
+		})
```
Comment (Devatoria) [inline container/docker_test.go]: Line jump between } and )

File: container/docker_test.go [hunk 5]
```diff
@@ -180,11 +213,49 @@ func (s *DockerContainerizerTestSuite) TestDockerContainerRemove() {
...
+	ips, err = s.dc.ContainerGetIPs("abcdef1234")
+	assert.Nil(s.T(), err)
+	assert.Equal(
+		s.T(),
+		map[string]net.IP{},
+		ips)
```
Comment (Devatoria) [inline container/docker_test.go]: Put the equal on the same line, or jump line before )

File: hook/iptables.go [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/Devatoria/go-mesos-executor/container"
+	"github.com/Devatoria/go-mesos-executor/logger"
+	"github.com/Devatoria/go-mesos-executor/types"
+	"github.com/coreos/go-iptables/iptables"
+	"github.com/spf13/viper"
+
```
Comment (Devatoria) [inline hook/iptables.go]: Empty line

File: hook/iptables.go [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+import (
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/Devatoria/go-mesos-executor/container"
+	"github.com/Devatoria/go-mesos-executor/logger"
+	"github.com/Devatoria/go-mesos-executor/types"
+	"github.com/coreos/go-iptables/iptables"
```
Comment (Devatoria) [inline hook/iptables.go]: Line jump before (to keep project internal packages and external packages split)

File: hook/iptables.go [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+			return err
+		}
+
+		driver, err := iptables.New()
+		if err != nil {
+			return err
+		}
+
+		// Go through all container IPs and ports to insert iptables
+
```
Comment (Devatoria) [inline hook/iptables.go]: Empty line

File: hook/iptables.go [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+			return err
+		}
+
+		driver, err := iptables.New()
+		if err != nil {
+			return err
+		}
+
+		for _, containerIP := range containerIPs {
+
```
Comment (Devatoria) [inline hook/iptables.go]: Empty line

File: hook/iptables.go [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+		if err != nil {
+			return err
+		}
+
+		driver, err := iptables.New()
+		if err != nil {
+			return err
+		}
+
+		for _, containerIP := range containerIPs {
```
Comment (Devatoria) [inline hook/iptables.go]: Please comment the loop

File: hook/iptables.go [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+		}
+
+		driver, err := iptables.New()
+		if err != nil {
+			return err
+		}
+
+		// Go through all container IPs and ports to insert iptables
+
+		for _, containerIP := range containerIPs {
```
Comment (Devatoria) [inline hook/iptables.go]: Comment the loop

File: hook/iptables_test.go [hunk 12]
```diff
@@ -0,0 +1,196 @@
...
+	monkey.PatchInstanceMethod(reflect.TypeOf(&isolated), "Close", func(_ *netns.NsHandle) error {
+		return nil
+	})
+
+	monkey.PatchInstanceMethod(
+		reflect.TypeOf(s.c),
+		"ContainerGetIPs",
+		func(_ *types.FakeContainerizer, id string) (map[string]net.IP, error) {
+			return s.containerIPs, nil
+		})
```
Comment (Devatoria) [inline hook/iptables_test.go]: Line jump between } and )

File: hook/iptables_test.go [hunk 13]
```diff
@@ -0,0 +1,196 @@
...
+		append(
+			BasePostRoutingRules,
+			[]string{
+				"-A POSTROUTING -s 172.0.2.1/32 -o !docker0 -j MASQUERADE",
+				"-A POSTROUTING -s 172.0.2.1/32 -d 172.0.2.1/32 -p tcp -m tcp --dport 80 -j MASQUERADE",
+				"-A POSTROUTING -s 172.0.2.1/32 -d 172.0.2.1/32 -p udp -m udp --dport 10000 -j MASQUERADE",
+				"-A POSTROUTING -s 172.0.3.1/32 -o !docker0 -j MASQUERADE",
+				"-A POSTROUTING -s 172.0.3.1/32 -d 172.0.3.1/32 -p tcp -m tcp --dport 80 -j MASQUERADE",
+				"-A POSTROUTING -s 172.0.3.1/32 -d 172.0.3.1/32 -p udp -m udp --dport 10000 -j MASQUERADE",
+			}...),
```
Comment (Devatoria) [inline hook/iptables_test.go]: Line jump between ... and )

File: hook/iptables_test.go [hunk 14]
```diff
@@ -0,0 +1,196 @@
...
+	assert.Subset(
+		s.T(),
+		append(
+			BaseForwardRule,
+			[]string{
+				"-A FORWARD -d 172.0.2.1/32 -i !docker0 -o docker0 -p tcp -m tcp --dport 80 -j ACCEPT",
+				"-A FORWARD -d 172.0.2.1/32 -i !docker0 -o docker0 -p udp -m udp --dport 10000 -j ACCEPT",
+				"-A FORWARD -d 172.0.3.1/32 -i !docker0 -o docker0 -p tcp -m tcp --dport 80 -j ACCEPT",
+				"-A FORWARD -d 172.0.3.1/32 -i !docker0 -o docker0 -p udp -m udp --dport 10000 -j ACCEPT",
+			}...),
```
Comment (Devatoria) [inline hook/iptables_test.go]: Line jump betweend ... and )

File: hook/iptables_test.go [hunk 15]
```diff
@@ -0,0 +1,196 @@
...
+			BasePostRoutingRules,
+			[]string{
+				"-A POSTROUTING -s 172.0.2.1/32 -o !docker0 -j MASQUERADE",
+				"-A POSTROUTING -s 172.0.2.1/32 -d 172.0.2.1/32 -p tcp -m tcp --dport 80 -j MASQUERADE",
+				"-A POSTROUTING -s 172.0.2.1/32 -d 172.0.2.1/32 -p udp -m udp --dport 10000 -j MASQUERADE",
+				"-A POSTROUTING -s 172.0.3.1/32 -o !docker0 -j MASQUERADE",
+				"-A POSTROUTING -s 172.0.3.1/32 -d 172.0.3.1/32 -p tcp -m tcp --dport 80 -j MASQUERADE",
+				"-A POSTROUTING -s 172.0.3.1/32 -d 172.0.3.1/32 -p udp -m udp --dport 10000 -j MASQUERADE",
+			}...),
+		postRoutingRules)
```
Comment (Devatoria) [inline hook/iptables_test.go]: Line jump before )

File: hook/iptables_test.go [hunk 16]
```diff
@@ -0,0 +1,196 @@
...
+		s.T(),
+		append(
+			BaseForwardRule,
+			[]string{
+				"-A FORWARD -d 172.0.2.1/32 -i !docker0 -o docker0 -p tcp -m tcp --dport 80 -j ACCEPT",
+				"-A FORWARD -d 172.0.2.1/32 -i !docker0 -o docker0 -p udp -m udp --dport 10000 -j ACCEPT",
+				"-A FORWARD -d 172.0.3.1/32 -i !docker0 -o docker0 -p tcp -m tcp --dport 80 -j ACCEPT",
+				"-A FORWARD -d 172.0.3.1/32 -i !docker0 -o docker0 -p udp -m udp --dport 10000 -j ACCEPT",
+			}...),
+		forwardRules)
```
Comment (Devatoria) [inline hook/iptables_test.go]: Line jump before )

File: hook/iptables_test.go [hunk 17]
```diff
@@ -0,0 +1,196 @@
...
+	assert.Subset(
+		s.T(),
+		append(
+			BasePreRoutingRules,
+			[]string{
+				"-A PREROUTING -i !docker0 -p tcp -m tcp --dport 32000 -j DNAT --to-destination 172.0.2.1:80",
+				"-A PREROUTING -i !docker0 -p udp -m udp --dport 33000 -j DNAT --to-destination 172.0.2.1:10000",
+				"-A PREROUTING -i !docker0 -p tcp -m tcp --dport 32000 -j DNAT --to-destination 172.0.3.1:80",
+				"-A PREROUTING -i !docker0 -p udp -m udp --dport 33000 -j DNAT --to-destination 172.0.3.1:10000",
+			}...),
```
Comment (Devatoria) [inline hook/iptables_test.go]: Line jump between ... and )

File: hook/iptables_test.go [hunk 18]
```diff
@@ -0,0 +1,196 @@
...
+		preRoutingRules)
```
Comment (Devatoria) [inline hook/iptables_test.go]: Line jump before )

File: config.yaml [hunk 19]
```diff
@@ -2,7 +2,10 @@ debug: true # Enable debug mode (more logs)
 hooks: # Enabled hooks
   - acl
   - removeContainer
+  - insertIptables
+  - removeIptables
 proc_path: /mnt/proc # Mount path of host proc folder
+container_bridge_interface: docker0
```
Comment (Devatoria) [inline config.yaml]: Maybe use embedded hook conf (for readability):

iptables:
  bridge_interface: docker0

File: hook/iptables.go [hunk 20]
```diff
@@ -0,0 +1,209 @@
...
+	},
+}
+
+// RemoveIptablesHook removes iptables injected by injectIptables hook.
+var RemoveIptablesHook = Hook{
+	Name:     "removeIptables",
+	Priority: 0,
+	Execute: func(c container.Containerizer, info *types.ContainerTaskInfo) error {
+		// Do not execute the hook if we are not on bridged network
+		if info.TaskInfo.GetContainer().GetDocker().GetNetwork() != mesos.ContainerInfo_DockerInfo_BRIDGE {
```
Comment (bewiwi) [inline hook/iptables.go:59]: Can you try to mutualise some code with InsterIptablesHook ? look likes a copy/paste with just little update

**Review by bewiwi — CHANGES_REQUESTED** [review]

File: hook/iptables.go [hunk 21]
```diff
@@ -0,0 +1,150 @@
...
+
+		GenerateIptables(c, info, driver.Delete)
+
+		return nil
+	},
+}
+
+// GenerateIptables generates all needed iptables for containers masquerading/ network forwarding.
+// The action function is called with each iptable generated.
+func GenerateIptables(c container.Containerizer, info *types.ContainerTaskInfo, action func(string, string, ...string) error) error {
```
Comment (Devatoria) [inline hook/iptables.go]: generateIptables (otherwise, this function will be exposed to other packages)

File: hook/iptables.go [hunk 22]
```diff
@@ -0,0 +1,150 @@
...
+		}
+
+		logger.GetInstance().Debug("Removing iptables on host namespace")
+
+		driver, err := iptables.New()
+		if err != nil {
+			return err
+		}
+
+		GenerateIptables(c, info, driver.Delete)
```
Comment (Devatoria) [inline hook/iptables.go]: The function returns an error but you do not check it. Please check it.

File: hook/iptables.go [hunk 23]
```diff
@@ -0,0 +1,150 @@
...
+		}
+
+		logger.GetInstance().Debug("Inserting iptables on host namespace")
+
+		driver, err := iptables.New()
+		if err != nil {
+			return err
+		}
+
+		GenerateIptables(c, info, driver.Append)
```
Comment (Devatoria) [inline hook/iptables.go]: The function returns an error but you do not check it. Please check it.

File: hook/iptables.go (hunk 21 above)
Comment (Devatoria) [reply hook/iptables.go]: Do not forget to edit the comment above once you changed the function name

File: hook/iptables.go (hunk 20 above)
Comment (WinterNis) [reply hook/iptables.go:59]: It is done

**Pull request merged**